dnf5-plugins/needs_restarting_plugin/needs_restarting.cpp

```diff
@@ -0,0 +1,316 @@
+    time_t uptime_boot_time = 0;
+    std::ifstream uptime_stream{"/proc/uptime"};
+    if (uptime_stream.is_open()) {
+        double uptime = 0;
+        uptime_stream >> uptime;
+        if (uptime > 0) {
+            uptime_boot_time = std::time(nullptr) - static_cast<time_t>(uptime);
+        }
+    }
+
+    const time_t boot_time = std::max(proc_1_boot_time, uptime_boot_time);
+
+    logger->debug("st_mtime of /proc/1: {}", proc_1_boot_time);
+    logger->debug("Boot time derived from /proc/uptime: {}", uptime_boot_time);
+    logger->debug("Using {} as the system boot time.", boot_time);
+    return boot_time;
+}
+
+libdnf5::rpm::PackageSet recursive_dependencies(
+    const libdnf5::rpm::Package & package, libdnf5::rpm::PackageQuery & installed) {
+    libdnf5::rpm::PackageSet dependencies{package.get_base()};
+    dependencies.add(package);
+
+    std::vector<libdnf5::rpm::Package> stack;
+    stack.emplace_back(package);
+
+    while (!stack.empty()) {
+        const auto & current = stack.back();
+        stack.pop_back();
+
+        libdnf5::rpm::PackageQuery query{installed};
+        query.filter_provides(current.get_requires());
+
+        for (const auto & dependency : query) {
+            if (!dependencies.contains(dependency)) {
+                stack.emplace_back(dependency);
+            }
+        }
+
+        dependencies |= query;
+    }
+
+    return dependencies;
+}
+
+void NeedsRestartingCommand::system_needs_restarting(Context & ctx) {
+    const auto boot_time = get_boot_time(ctx);
+
+    libdnf5::rpm::PackageQuery reboot_suggested{ctx.base};
+    reboot_suggested.filter_installed();
+    reboot_suggested.filter_reboot_suggested();
+
+    std::vector<libdnf5::rpm::Package> need_reboot = {};
+    for (const auto & pkg : reboot_suggested) {
+        if (pkg.get_install_time() > static_cast<unsigned long long>(boot_time)) {
+            need_reboot.push_back(pkg);
+        }
+    }
+
+    if (need_reboot.empty()) {
+        std::cout << "No core libraries or services have been updated since boot-up." << std::endl
+                  << "Reboot should not be necessary." << std::endl;
+    } else {
+        std::cout << "Core libraries or services have been updated since boot-up:" << std::endl;
+        for (const auto & pkg : need_reboot) {
+            std::cout << "  * " << pkg.get_name() << std::endl;
+        }
+        std::cout << std::endl
+                  << "Reboot is required to fully utilize these updates." << std::endl
+                  << "More information: https://access.redhat.com/solutions/27943" << std::endl;
+        throw libdnf5::cli::SilentCommandExitError(1);
+    }
+}
+
+void NeedsRestartingCommand::services_need_restarting(Context & ctx) {
+    std::unique_ptr<sdbus::IConnection> connection;
+    try {
+        connection = sdbus::createSystemBusConnection();
+    } catch (const sdbus::Error & ex) {
+        throw libdnf5::cli::CommandExitError(1, M_("Couldn't connect to D-Bus: {}"), ex.what());
```

Comment:
Copr build is failing for now. There seems to be some issue when passing the given arguments here to the underlying Error classes, although at first glance, I don't see why...

Reply:
OK, the Copr build is succeeding now after creating a `std::string` from `ex.what()`. Not sure why it only failed in Copr when passing the `char *`.